firedrake/rf_eigensolver.py

```diff
@@ -0,0 +1,661 @@
+    additionally able to cope with eigenproblems arising from
+    linearisations of operators represented by sequences of
+    Firedrake operations.
+
+    Parameters
+    ----------
+    problem : 
+        The eigenproblem to solve.
+    n_evals : 
+        The number of eigenvalues to compute.
+    options_prefix : 
+        The options prefix to use for the eigensolver.
+    solver_parameters : 
+        PETSc options for the eigenvalue problem.
+    ncv : 
+        Maximum dimension of the subspace to be used by the solver. See
+        `SLEPc.EPS.setDimensions`.
+    mpd : 
+        Maximum dimension allowed for the projected problem. See
+        `SLEPc.EPS.setDimensions`.
+
+    Notes
+    -----
+    Users will typically wish to set solver parameters specifying the symmetry
+    of the eigenproblem and which eigenvalues to search for first.
+
+    The former is set using the options available for `EPSSetProblemType
+    <https://slepc.upv.es/documentation/current/docs/manualpages/EPS/EPSSetProblemType.html>`__.
+
+    For example if the bilinear form is symmetric (Hermitian in complex mode),
+    one would add this entry to `solver_options`::
+
+        "eps_gen_hermitian": None
+    
+    Note that for eigenproblems arising from taped linearised operators, matrices in the
+    formulation are implicit (`PETSc.Mat.PYTHON` type), which restricts the available 
+    options to those that do not use the explicit matrix entries. For instance, spectral 
+    transforms won't work and solvers are restricted. 
+
+    As always when specifying PETSc options, `None` indicates that the option
+    in question is a flag and hence doesn't take an argument.
+
+    The eigenvalues to search for first are specified using the options for
+    `EPSSetWhichEigenPairs <https://slepc.upv.es/documentation/current/docs/manualpages/EPS/EPSSetWhichEigenpairs.html>`__.
+
+    For example, to look for the eigenvalues with largest real part, one
+    would add this entry to `solver_options`::
+
+        "eps_largest_real": None
+    """
+
+    DEFAULT_EPS_PARAMETERS = {
+        "eps_type": "krylovschur",
+        "eps_tol": 1e-10,
+        "eps_target": 0.0,
+    }
+
+    def __init__(
+        self,
+        problem: RFEigenproblem,
+        n_evals: int,
+        *,
+        options_prefix: str = None,
+        solver_parameters: dict | None = None,
+        ncv: int = None,
+        mpd: int = None,
+    ):
+        self.es = SLEPc.EPS().create(comm=problem.dm.comm)
+        self._problem = problem
+        self.n_evals = n_evals
+        self.ncv = ncv
+        self.mpd = mpd
+        solver_parameters = flatten_parameters(solver_parameters or {})
+        for key in self.DEFAULT_EPS_PARAMETERS:
+            value = self.DEFAULT_EPS_PARAMETERS[key]
+            solver_parameters.setdefault(key, value)
+        self.options_manager = OptionsManager(solver_parameters, options_prefix)
+        self.options_manager.set_from_options(self.es)
```

Comment:
This is now the recommended way to handle solver options.
You'll need to import `attach_options`, `set_from_options`, `set_default_parameter`, and `inserted_options` from `petsctools`.
```suggestion
        attach_options(self.es, solver_parameters or {}, options_prefix)
        for key, value in self.DEFAULT_EPS_PARAMETERS:
            set_default_parameter(self.es, key, val)
        set_from_options(self.es)        
```